### DISCUSSION TITLE
NextJS Vercel fetch failed

### DISCUSSION THREAD

**Discussion by FedkoAnastasiia** [vercel/vercel#11626] [Help]
### Question

I get an error when deploying NextJS project on Vercel: 

<img width="889" alt="Screenshot 2024-05-22 at 21 12 34" src="https://github.com/vercel/vercel/assets/96294351/32e640d6-13b4-454d-b922-6d9abf193586">

Here is the structure of my app folder:
app
- events
   - page.tsx
- blog
   - page.tsx
- api
  - content
     - route.ts
  - events
     - route.ts
- layout.tsx
- page.tsx

I am fetching data from api folder and it works on localhost but not on Vercel.

This is my code:

- app/api/events/route.ts
```
import { IEvent } from "@/common/interfaces";
import { NextResponse } from "next/server";

const events: IEvent[] = [
 // events
];

export async function GET() {
  return NextResponse.json(events);
}
```

- app/events/page.tsx

```
import { IEvent } from "@/common/interfaces";
import EventPrewiew from "@/components/EventPrewiew";

const getEvents = async () => {
  return fetch(`${process.env.URL}/api/events`, {
    method: "GET",
    headers: {
      Accept: "application/json",
    },
  })
    .then((res) => res.json())
    .catch((err) => console.log(err.message));
};

const Events = async () => {
  const events: IEvent[] = await getEvents();

  console.log("events", events);

  return (
    <div>
      {events.map((e) => (
        <EventPrewiew key={e.id} event={e} />
      ))}
    </div>
  );
};

export default Events;
```

- .env.local 
`URL=http://localhost:3000`

Vercel project environment variable: 
`URL - https://my-project.vercel.app`

Please help to fix this and understand why it works locally and not on Vercel.

**Comment by math-silva** [comment]
Absolutely! Here are two suggestions to resolve this Vercel deployment issue:

**Suggestion 1: Update the Environment Variable Name**

* **Change:** Modify your environment variable name from `URL` to `NEXT_PUBLIC_URL`.

* **Reason:** In Next.js, environment variables prefixed with `NEXT_PUBLIC_` are embedded into the client-side bundle. This means your frontend code in `events/page.tsx` can access it correctly during both development and production on Vercel.

**Suggestion 2: Use the Standard `Response` Object**

* **Change:** Replace `NextResponse.json(events)` with `Response.json(events)`.

* **Reason:**  While `NextResponse` offers convenient helpers, the standard `Response` object is more universally compatible and ensures consistent behavior across different environments. 

**Modified Code:**

```typescript
// app/events/page.tsx
// ... (rest of the code)

const getEvents = async () => {
  return fetch(`${process.env.NEXT_PUBLIC_URL}/api/events`, { 
    // ... (rest of the code)
  });
};
```

**Updated Vercel Environment Variables:**

```
NEXT_PUBLIC_URL - https://my-project.vercel.app
```

Let me know if this solution helps you resolve the issue! If it does, please consider marking this answer as accepted.

**Reply by FedkoAnastasiia** [reply]
I did everything as you said, but unfortunately, it didn't help. 
Any other ideas on how to fix this?
